feat(guides): add new guide create-gasless-nft-drop (#63)
nftlabs/portal: portal/guides/create-gasless-nft-drop.mdx

---
title: How to create a gasless NFT drop with Next.js and thirdweb
author: Avneesh Agarwal
authorAvatar: "/assets/portal/authors/avneesh.jpg"
authorTwitter: "https://twitter.com/avneesh0612"
draft: false
tags: [edition-drop, react, claim, gas]

> summary: "Create a gasless NFT drop with Next.js and thirdweb"

Reviewer: this should be a better explainer and subtitle of what the guide would be about, my suggestion:

"Learn how to create an NFT Drop in which you pay the gas for your users, using OpenZeppeling, Next.js and thirdweb"

image: "/assets/portal/guides/create-nft-drop-with-claim-phases/thumbnail.png"
date: "2022-04-03"
updated: ""
---

> In some cases, you might wanna pay the gas for NFT drops like early access drops or when you are accepting payments in some other forms. So in this guide, we are going to see how to create a gasless NFT drop that everyone can claim. I am going to use an ERC-1155 so multiple people can hold the same NFT but you can do it any way you like!

Reviewer: would be good to also link to the finished code here

And select Edition drop. If you want to drop unique NFTs you can select a normal drop

Fill out the details and hit deploy!

![New Edition drop contract](/assets/portal/guides/create-gasless-nft-drop/new-edition-drop-contract.png)

After the drop is created, create a new NFT inside of the drop like this-

![New Editon drop](/assets/portal/guides/create-gasless-nft-drop/new-edition-drop.png)

Finally, let’s add a claim phase! Click on the settings button for the token, update the details based on what you need, and hit update. You will be asked to confirm a small transaction so confirm that.

![Claim Phase](/assets/portal/guides/create-gasless-nft-drop/claim-phase.png)

## Building the website

We will now build a website where users will be able to come and claim the NFT!

### Setup

Create a new Next.js app-

```bash
npx create-next-app gasless-drop
```

Install the dependencies we need-

```bash
npm i @thirdweb-dev/react @thirdweb-dev/sdk ethers # npm

yarn add @thirdweb-dev/react @thirdweb-dev/sdk ethers # yarn
```

### Adding metamask auth

We are also going to need authentication in the app because users will need to sign a message as well!

> Wrap the whole app in the thirdweb Provider like the following-

Reviewer: ```suggestion
Wrap the whole app in the `ThirdwebProvider` like the following:
```

```

We will create a new function to claim the NFT-

```jsx
const tokenId = 0;
const quantity = 1;

const claimNFT = async () => {
  try {
    await editionDrop?.claimTo(address, tokenId, quantity);
    console.log("🎉 NFT claimed successfully!");
  } catch (err) {
    console.log("💩 Error claiming NFT: ", err);
  }
};
```

The edition drop has a method called `claimTo` that we can use to claim the NFT to a specific wallet address. I have created two variables for the tokenId and quantity for better readability of code. The second arg is tokenId and the third is quantity as it can be seen.

Let’s now try minting the NFT!

<div style={{position: "relative", paddingBottom: "56.25%", height: 0}}><iframe src="https://www.loom.com/embed/7a1d9a37c90e4829abe7887b490ab006" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen
style={{position: "absolute", top: "0",left: "0", height: "100%", width: "100%"}}
></iframe></div>

And the minting works! But we have to pay for the gas in the transaction, so let’s relay the gas so that we pay the gas and user can mint their NFTs for free!

## Making transactions gasless

> We are going to use [Openzeppelin](https://defender.openzeppelin.com/) to relay the gas, so head over to Openzeppelin and sign up for an account if you haven’t already.

Reviewer: would be good to mention that thirdweb also supports Biconomy

> Create a new relayer in openzepplin-

Reviewer: ```suggestion
Create a new relayer in OpenZeppelin
```

> Create a new relayer in OpenZeppelin-

Reviewer: ```suggestion
Create a new relayer in OpenZeppelin:
```

Can we use colons (:) instead of dashes (-) for these types of things?

You should be able to see the amount you transferred in your dashboard

![Relayer Dashboard](/assets/portal/guides/create-gasless-nft-drop/relayer-dashboard.png)

Now go ahead and create a new autotask, you will be asked to fill some data-

![Create Autotask](/assets/portal/guides/create-gasless-nft-drop/create-autotask.png)

You can get the code from [here](https://raw.githubusercontent.com/thirdweb-dev/ozdefender-autotask/main/src/forwarder_handler.js) and paste it into the code section of autotask. After you are done hit create. Once it is done creating, copy the webhook uri

![Copy webhook uri](/assets/portal/guides/create-gasless-nft-drop/copy-webhook-uri.png)

Head back to the Next.js application, and create a new `.env.local` file in the root of the folder. We will use this to store our environment variables! Create a new variable `NEXT_PUBLIC_OPENZEPPELIN_URL` like this-

```jsx
NEXT_PUBLIC_OPENZEPPELIN_URL=https://api.defender.openzeppelin.com/autotasks/secret-stuff
```

Since we have changed env variables, you need to restart your development server.

> Finally, add the OpenZeppelin Uri in the thirdwebProvider-

Reviewer: ```suggestion
Finally, add the OpenZeppelin Uri in the `ThirdwebProvider`:
```

```jsx
<ThirdwebProvider
  sdkOptions={{
    gasless: {
      openzeppelin: {
        relayerUrl: process.env.NEXT_PUBLIC_OPENZEPPELIN_URL,
      },
    },
  }}
  desiredChainId={desiredChainId}
>
```

Now the gas fees is being relayed and the users don’t have to pay any gas!

<div style={{position: "relative", paddingBottom: "56.25%", height: 0}}><iframe src="https://www.loom.com/embed/0b1a961b538146678c29a1fa26f17799" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen
style={{position: "absolute", top: "0",left: "0", height: "100%", width: "100%"}}
></iframe></div>

> Want to have a look at the code? [We got you!](https://github.com/avneesh0612/gasless-drop)

Reviewer: examples repo PR is merged, change the link here to the folder there
